File: src-tui/scanner.ts
import { spawn, spawnSync } from "node:child_process";
import { existsSync, readFileSync, readlinkSync } from "node:fs";
import { basename, dirname, join } from "node:path";
import type { PortInfo } from "./types";

const PROJECT_MARKERS = [
  "package.json",
  "Cargo.toml",
  "go.mod",
  "pyproject.toml",
  "requirements.txt",
  "pom.xml",
  "build.gradle",
  ".git",
];

function sleepMs(ms: number): void {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
}

function run(cmd: string, args: string[]): string {
  const result = spawnSync(cmd, args, {
    encoding: "utf8",
    maxBuffer: 10 * 1024 * 1024,
  });
  if (result.error) throw result.error;
  return result.stdout ?? "";
}

function findProjectRoot(start: string | null): string | null {
  if (!start) return null;
  let dir = start;
  for (let i = 0; i < 6; i++) {
    for (const marker of PROJECT_MARKERS) {
      if (existsSync(join(dir, marker))) return dir;
    }
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

function extractProjectName(path: string | null): string | null {
  if (!path) return null;
  const name = basename(path);
  return name || null;
}

function getProcessMetaUnix(pid: number): {
  process_name: string | null;
  project_path: string | null;
  start_cmd: string | null;
} {
  let process_name: string | null = null;
  let project_path: string | null = null;
  let start_cmd: string | null = null;

  try {
    const cmdline = readFileSync(`/proc/${pid}/cmdline`, "utf8");
    if (cmdline) {
      start_cmd = cmdline.replace(/\0/g, " ").trim() || null;
    }
  } catch {
    // ignore
  }

  try {
    const cwd = readlinkSync(`/proc/${pid}/cwd`);
    project_path = findProjectRoot(cwd);
  } catch {
    // ignore
  }

  try {
    const comm = readFileSync(`/proc/${pid}/comm`, "utf8").trim();
    if (comm) process_name = comm;
  } catch {
    // ignore
  }

  if (!project_path && process.platform === "darwin") {
    try {
      const out = run("lsof", ["-p", String(pid), "-a", "-d", "cwd", "-Fn"]);
      const line = out.split("\n").find((l) => l.startsWith("n") && l.length > 1);
      if (line) {
        project_path = findProjectRoot(line.slice(1));
      }
    } catch {
      // ignore
    }
  }

  if (!process_name) {
    try {
      const out = run("ps", ["-p", String(pid), "-o", "comm="]);
      const name = out.trim();
      if (name) process_name = name;
    } catch {
      // ignore
    }
  }

  if (!start_cmd) {
    try {
      const out = run("ps", ["-p", String(pid), "-o", "args="]);
      const args = out.trim();
      if (args) start_cmd = args;
    } catch {
      // ignore
    }
  }

  return { process_name, project_path, start_cmd };
}

function scanUnix(): PortInfo[] {
  const stdout = run("lsof", ["-iTCP", "-sTCP:LISTEN", "-n", "-P"]);
  const ports: PortInfo[] = [];
  const seen = new Set<string>();

  for (const line of stdout.split("\n").slice(1)) {
    const parts = line.trim().split(/\s+/);
    if (parts.length < 9) continue;

    let process_name = parts[0];
    const pid = Number(parts[1]);
    if (!Number.isFinite(pid) || pid <= 0) continue;

    // NAME may be "*:3000", "127.0.0.1:3000", or "127.0.0.1:3000 (LISTEN)"
    const addrPart =
      parts.find((p) => /:\d+$/.test(p)) ??
      parts.find((p) => p.includes(":") && /\d/.test(p));
    if (!addrPart) continue;
    const portMatch = addrPart.match(/:(\d+)\s*$/);
    const port = portMatch ? Number(portMatch[1]) : NaN;
    if (!Number.isFinite(port)) continue;

    const key = `${port}:${pid}`;
    if (seen.has(key)) continue;
    seen.add(key);

    const meta = getProcessMetaUnix(pid);
    if (meta.process_name) process_name = meta.process_name;

    ports.push({
      port,
      pid,
      process_name,
      project_path: meta.project_path,
      project_name: extractProjectName(meta.project_path),
      protocol: "TCP",
      start_cmd: meta.start_cmd,
    });
  }

  ports.sort((a, b) => a.port - b.port);
  return ports;
}

function scanWindows(): PortInfo[] {
  const stdout = run("netstat", ["-ano"]);
  const ports: PortInfo[] = [];
  const seen = new Set<string>();

  for (const line of stdout.split("\n")) {
    if (!line.includes("LISTENING")) continue;
    const parts = line.trim().split(/\s+/);
    if (parts.length < 5) continue;

    const local = parts[1];
    const portStr = local.includes(":") ? local.split(":").pop() : null;
    const port = portStr ? Number(portStr) : NaN;
    const pid = Number(parts[parts.length - 1]);
    if (!Number.isFinite(port) || !Number.isFinite(pid) || pid <= 0) continue;

    const key = `${port}:${pid}`;
    if (seen.has(key)) continue;
    seen.add(key);

    let process_name = `PID ${pid}`;
    let start_cmd: string | null = null;
    try {
      const out = run("tasklist", ["/FI", `PID eq ${pid}`, "/FO", "CSV", "/NH"]);
      const m = out.match(/"([^"]+)","(\d+)"/);
      if (m) process_name = m[1];
    } catch {
      // ignore
    }

    try {
      const out = run("wmic", [
        "process",
        "where",
        `ProcessId=${pid}`,
        "get",
        "CommandLine",
        "/value",
      ]);
      const cmdLine = out
        .split("\n")
        .find((l) => l.startsWith("CommandLine="))
        ?.slice("CommandLine=".length)
        .trim();
      if (cmdLine) start_cmd = cmdLine;
    } catch {
      // ignore
    }

    ports.push({
      port,
      pid,
      process_name,
      project_path: null,
      project_name: null,
      protocol: "TCP",
      start_cmd,
    });
  }

  ports.sort((a, b) => a.port - b.port);
  return ports;
}

/** Scan listening TCP ports (mirrors the Rust scanner logic). */
export function scanPorts(): PortInfo[] {
  if (process.platform === "win32") return scanWindows();
  return scanUnix();
}

export function killProcess(pid: number): void {
  if (process.platform === "win32") {
    const result = spawnSync("taskkill", ["/PID", String(pid), "/F"], {
      encoding: "utf8",
    });
    if (result.status !== 0) {
      throw new Error(result.stderr || `Failed to kill PID ${pid}`);
    }
    return;
  }

  try {
    process.kill(pid, "SIGTERM");
  } catch (e) {
    throw new Error(`Failed to kill PID ${pid}: ${e}`);
  }

  const deadline = Date.now() + 2000;
  while (Date.now() < deadline) {
    try {
      process.kill(pid, 0);
      sleepMs(50);
    } catch {
      return;
    }
  }

  try {
    process.kill(pid, "SIGKILL");
  } catch {
    // already dead
  }
}

export function restartProcess(pid: number, cmd: string, cwd: string): void {
  killProcess(pid);
  sleepMs(800);

  if (process.platform === "win32") {
    const child = spawn("cmd", ["/C", "start", "cmd", "/K", cmd], {
      cwd,
      detached: true,
      stdio: "ignore",
    });
    child.unref();
    return;
  }

  const parts = cmd.split(/\s+/).filter(Boolean);
  const program = parts[0];
  if (!program) throw new Error("empty command");
  const args = parts.slice(1);

  const child = spawn(program, args, {
    cwd,
    detached: true,
    stdio: "ignore",
  });
  child.unref();
}

File: src-tauri/src/scanner.rs
use serde::Serialize;
use std::collections::HashSet;
use std::process::Command;
use sysinfo::System;

#[derive(Serialize, Clone)]
pub struct PortInfo {
    pub port: u16,
    pub pid: u32,
    pub process_name: String,
    pub project_path: Option<String>,
    pub project_name: Option<String>,
    pub protocol: String,
    pub start_cmd: Option<String>,
}

// ─── Entry point (platform router) ───────────────────────────────────────────

pub fn scan_ports() -> Vec<PortInfo> {
    let mut sys = System::new();
    sys.refresh_processes();

    #[cfg(target_os = "windows")]
    return scan_windows(&sys);

    #[cfg(any(target_os = "macos", target_os = "linux"))]
    return scan_unix(&sys);
}

pub fn kill_pid(pid: u32) -> Result<(), String> {
    #[cfg(target_os = "windows")]
    return kill_windows(pid);

    #[cfg(any(target_os = "macos", target_os = "linux"))]
    return kill_unix(pid);
}

// ─── Windows ─────────────────────────────────────────────────────────────────

#[cfg(target_os = "windows")]
fn scan_windows(sys: &System) -> Vec<PortInfo> {
    let output = Command::new("netstat")
        .args(["-ano"])
        .output()
        .expect("failed to run netstat");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let mut ports: Vec<PortInfo> = Vec::new();
    let mut seen_entries: HashSet<(u16, u32)> = HashSet::new();

    for line in stdout.lines() {
        if !line.contains("LISTENING") {
            continue;
        }
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 5 {
            continue;
        }

        let port: u16 = match parts[1].rsplit(':').next().and_then(|p| p.parse().ok()) {
            Some(p) => p,
            None => continue,
        };

        let pid: u32 = match parts[4].parse() {
            Ok(p) => p,
            Err(_) => continue,
        };

        if pid == 0 || seen_entries.contains(&(port, pid)) {
            continue;
        }
        seen_entries.insert((port, pid));

        let mut process_name = format!("PID {}", pid);
        let mut project_path = None;
        let mut start_cmd = None;

        if let Some(process) = sys.process(sysinfo::Pid::from(pid as usize)) {
            let p_name = process.name().to_string();
            if !p_name.trim().is_empty() {
                process_name = p_name;
            }

            let cmd_arr = process.cmd();
            let cmd_str = cmd_arr.join(" ");

            if !cmd_str.trim().is_empty() {
                start_cmd = Some(cmd_str.trim().to_string());
            }

            if let Some(cwd) = process.cwd() {
                project_path = find_project_root(cwd).map(|p| p.to_string_lossy().to_string());
            }

            if project_path.is_none() {
                if let Some(exe) = process.exe() {
                    if let Some(parent) = exe.parent() {
                        project_path =
                            find_project_root(parent).map(|p| p.to_string_lossy().to_string());
                    }
                }
            }
        }

        let project_name = extract_project_name(&project_path);

        ports.push(PortInfo {
            port,
            pid,
            process_name,
            project_path,
            project_name,
            protocol: "TCP".into(),
            start_cmd,
        });
    }

    ports.sort_by_key(|p| p.port);
    ports
}

#[cfg(target_os = "windows")]
fn kill_windows(pid: u32) -> Result<(), String> {
    let output = Command::new("taskkill")
        .args(["/PID", &pid.to_string(), "/F"])
        .output()
        .map_err(|e| e.to_string())?;

    if output.status.success() {
        Ok(())
    } else {
        Err(String::from_utf8_lossy(&output.stderr).to_string())
    }
}

// ─── macOS + Linux (shared lsof path) ────────────────────────────────────────

#[cfg(any(target_os = "macos", target_os = "linux"))]
fn scan_unix(sys: &System) -> Vec<PortInfo> {
    let output = Command::new("lsof")
        .args(["-iTCP", "-sTCP:LISTEN", "-n", "-P"])
        .output()
        .expect("failed to run lsof — is it installed?");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let mut ports: Vec<PortInfo> = Vec::new();
    let mut seen_entries: HashSet<(u16, u32)> = HashSet::new();

    for line in stdout.lines().skip(1) {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 9 {
            continue;
        }

        let mut process_name = parts[0].to_string();
        let pid: u32 = match parts[1].parse() {
            Ok(p) => p,
            Err(_) => continue,
        };

        let name = parts[parts.len() - 1];
        let port: u16 = match name.rsplit(':').next().and_then(|p| p.parse().ok()) {
            Some(p) => p,
            None => continue,
        };

        if seen_entries.contains(&(port, pid)) {
            continue;
        }
        seen_entries.insert((port, pid));

        let mut project_path = None;
        let mut start_cmd = None;

        if let Some(process) = sys.process(sysinfo::Pid::from(pid as usize)) {
            let sys_name = process.name().to_string();
            if !sys_name.is_empty() {
                process_name = sys_name;
            }

            if let Some(cwd) = process.cwd() {
                project_path = find_project_root(cwd).map(|p| p.to_string_lossy().to_string());
            }

            let cmd = process.cmd().join(" ");
            if !cmd.trim().is_empty() {
                start_cmd = Some(cmd);
            }
        }

        if project_path.is_none() {
            project_path = get_project_path_unix_fallback(pid);
        }

        let project_name = extract_project_name(&project_path);

        ports.push(PortInfo {
            port,
            pid,
            process_name,
            project_path,
            project_name,
            protocol: "TCP".into(),
            start_cmd,
        });
    }

    ports.sort_by_key(|p| p.port);
    ports
}

#[cfg(any(target_os = "macos", target_os = "linux"))]
fn get_project_path_unix_fallback(pid: u32) -> Option<String> {
    #[cfg(target_os = "linux")]
    {
        let cwd = std::fs::read_link(format!("/proc/{}/cwd", pid)).ok()?;
        find_project_root(std::path::Path::new(&cwd)).map(|p| p.to_string_lossy().to_string())
    }

    #[cfg(target_os = "macos")]
    {
        let output = Command::new("lsof")
            .args(["-p", &pid.to_string(), "-a", "-d", "cwd", "-Fn"])
            .output()
            .ok()?;

        let s = String::from_utf8_lossy(&output.stdout);
        let cwd = s
            .lines()
            .find(|l| l.starts_with('n') && l.len() > 1)
            .map(|l| l[1..].to_string())?;

        find_project_root(std::path::Path::new(&cwd)).map(|p| p.to_string_lossy().to_string())
    }
}

#[cfg(any(target_os = "macos", target_os = "linux"))]
fn kill_unix(pid: u32) -> Result<(), String> {
    unsafe {
        libc::kill(pid as i32, libc::SIGTERM);
    }
    std::thread::sleep(std::time::Duration::from_secs(2));

    if process_exists_unix(pid) {
        unsafe {
            libc::kill(pid as i32, libc::SIGKILL);
        }
    }
    Ok(())
}

#[cfg(any(target_os = "macos", target_os = "linux"))]
fn process_exists_unix(pid: u32) -> bool {
    unsafe { libc::kill(pid as i32, 0) == 0 }
}

// ─── Shared helpers ───────────────────────────────────────────────────────────

fn find_project_root(start: &std::path::Path) -> Option<std::path::PathBuf> {
    let markers = [
        "package.json",
        "Cargo.toml",
        "go.mod",
        "pyproject.toml",
        "requirements.txt",
        "pom.xml",
        "build.gradle",
        ".git",
    ];
    let mut dir = start.to_path_buf();
    for _ in 0..6 {
        for marker in &markers {
            if dir.join(marker).exists() {
                return Some(dir);
            }
        }
        if !dir.pop() {
            break;
        }
    }
    None
}

fn extract_project_name(path: &Option<String>) -> Option<String> {
    path.as_ref()
        .and_then(|p| std::path::Path::new(p).file_name())
        .and_then(|n| n.to_str())
        .map(|s| s.to_string())
}
